### PR TITLE
Add origins check for websockets

### PR DESCRIPTION
**Pull request by asvetlov**
I would like to add check for "Origin" header.
Do you like the idea?
I'll add required tests to PR, sure.

### CODE REVIEW

File: aiohttp/websocket.py [hunk 1]
```diff
@@ -226,6 +226,14 @@ def do_handshake(method, headers, transport, protocols=()):
                 'Client protocols %r don’t overlap server-known ones %r',
                 protocols, req_protocols)
 
+    if origins:
+        origin = headers.get('ORIGIN', '').strip()
+        if origin not in origins:
```
Comment (kxepal) [inline aiohttp/websocket.py:231]: [It doesn't simply](http://tools.ietf.org/html/rfc6454#section-4) to just search origin header value in some list seeking for [comparison](http://tools.ietf.org/html/rfc6454#section-5):
0. Origin value should be split into scheme, host, port triple
1. Host need to be converted to lower case
2. If port is not specified, the default one should be picked from the specified scheme
3. And only now compare scheme/host/port triples

**Comment by ludovic-gasc** [comment]
I'm not a big fan, you can't really trust a client based on Web browser security. 
But, if it's not mandatory, it's not a problem. 

Could you change the logging level about no origins from warn to info? 
Thank you.

File: aiohttp/websocket.py (hunk 1 above)
Comment (asvetlov) [inline aiohttp/websocket.py:231]: Good point

**Comment by kxepal** [comment]
@GMLudo CORS isn't web browsers security, but web browsers are the most active users of this technique. If aiohttp.web tends to serve content for the world in public, it need to follow common HTTP standards.

**Comment by asvetlov** [comment]
Ok, reducing logger level makes sense.

**Comment by asvetlov** [comment]
Let me make a pause. Maybe CORS middleware should handle Origin?

**Comment by kxepal** [comment]
@asvetlov imho, yes, since Origin header comes from CORS specification.

**Comment by fafhrd91** [comment]
Agree, origin should be part of cors. But I think cors should be enforced
on endpoint level, rather on middleware level

**Comment by asvetlov** [comment]
Sure, cors should be enabled per route.
I'd like to present aiohttp_cors, please wait for a while.

**Comment by fafhrd91** [comment]
Route could accept middlewares same as application.

**Comment by bluetech** [comment]
I browsed this library, and noticed this issue. I think you should revisit your decision to close this.

I don't think it makes much sense to defer this to a CORS library, since CORS is used to relax single-origin-policy, which does not apply to websockets (they allow all origins by default).

I think that for security related knobs, the default should be safe, and people who "know what they are doing" can change them.  This article explains why I consider the current behaviour to be a bad default: https://www.christian-schneider.net/CrossSiteWebSocketHijacking.html

I like the way Tornado handles this (http://www.tornadoweb.org/en/stable/websocket.html#tornado.websocket.WebSocketHandler.check_origin) - there's a `check_origin` callback, which is called with the origin if present, and by default compares Origin to Host and rejects if different. So the default is safe, but overridable per endpoint.

If you do not like to implement this, then it should at least be noted in the documentation, because many/most developers would not consider it otherwise.

Thanks for considering - do note that I am not an expert about this, so take with a grain of salt.

**Comment by asvetlov** [comment]
@rutsky that do you think about?
From my perspective we may add a callback to aiohttp_cors for specific routes at least.

**Comment by rutsky** [comment]
1. `Origin` header check on server side in WebSockets is optional, so `aiohttp` is not obligated to implement it.
2. WebSockets are not subject of Same Origin Policy, so technically CORS is unrelated here. Browser will always try to initiate WS connection and [will](https://tools.ietf.org/html/rfc6455#section-4.2.1) send proper `Origin` header (for other clients this header is optional), so [it's up to server](https://tools.ietf.org/html/rfc6455#section-4.2.2) to allow or deny WS connection.
3. Mentioned by @bluetech Web Socket Hijacking is a real threat and IMO it's better to check `Origin` in `aiohttp` by default, rather than defer this to the user. See also https://tools.ietf.org/html/rfc6455#page-50

So I think `Origin` check need to be implemented in `aiohttp`.

I like the default behaviour of Tornado, mentioned by @bluetech, and it doesn't require any additional configuration data.

To properly check origin additional steps required, as mentioned by @kxepal.
These steps are not implemented in `aiohttp_cors`, because CORS specification allows to compare origins simply as strings, see https://www.w3.org/TR/cors/#resource-requests:

> If the value of the Origin header is not a case-sensitive match for any of the values in `list of origins` do not set any additional headers and terminate this set of steps.
> Note: Always matching is acceptable since the list of origins can be unbounded.

In `aiohttp_cors` it's assumed, that user in CORS configuration specifies `list of origins`,
and only check should be done that origin is meaningful string (here is TODO for it: https://github.com/aio-libs/aiohttp_cors/blob/master/aiohttp_cors/cors_config.py#L73).

I think origin normalization/comparison should be implemented in `aiohttp`, and then we may be use it in `aiohttp_cors`, so I still don't think dependency on `aiohttp_cors` is required here.

**Pull request closed without merging**